### PR TITLE
transform: adds url_decode keyword

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2689

Describe changes:
- Adds url_decode transform keyword (and even updates the doc)

There are some open questions :
- should this PR limit the size of the input buffer ? as mentioned here https://redmine.openinfosecfoundation.org/issues/2689#note-1
- What should be the behavior in the case the transform fails ? (that is the buffer is not a valid url-encoded one and has patterns such as `%ZZ`) Right, now I keep undecoded data (like https://www.urldecoder.org for instance)
- The use of a transform looked good to me as it can get reused and is only computed if needed, but there is the suggestion to do it like `http_uri`

There is no reuse of libhtp code as libhtp `htp_decode_path_inplace` mixes more context, such as setting `response_status_expected_number` for the transaction...

Makes test https://github.com/OISF/suricata-verify/pull/146 pass

Modifies #4347 by fixing transformed buffer size and avoiding overflow

### CODE REVIEW

File: src/detect-transform-urldecode.c [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Philippe Antoine <p.antoine@catenacyber.fr>
+ *
+ * Implements the urldecode transform keyword
```
Comment (jlucovsky) [inline src/detect-transform-urldecode.c]: nit: s/urldecode/url_decode/

File: src/detect-http-client-body.c [hunk 2]
```diff
@@ -258,6 +258,7 @@ static InspectionBuffer *HttpClientBodyGetDataCallback(DetectEngineThreadCtx *de
     StreamingBufferGetDataAtOffset(body->sb,
             &data, &data_len, offset);
     InspectionBufferSetup(buffer, data, data_len);
+    InspectionBufferApplyTransforms(buffer, transforms);
```
Comment (victorjulien) [inline src/detect-http-client-body.c]: this should go into a separate commit as it is not related to urldecode

File: src/detect-transform-urldecode.c [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+ */
+static int DetectTransformUrlDecodeSetup (DetectEngineCtx *de_ctx, Signature *s, const char *nullstr)
+{
+    SCEnter();
+    int r = DetectSignatureAddTransform(s, DETECT_TRANSFORM_URL_DECODE);
+    SCReturnInt(r);
+}
+
+// util function so as to ease reuse sometimes
+static uint32_t utilBufferUrlDecode(const uint8_t *input, const uint32_t input_len, uint8_t *output)
```
Comment (victorjulien) [inline src/detect-transform-urldecode.c]: style: proper name would be ``UtilBufferUrlDecode`` (starting with capital U) but I prefer just dropping the ``util`` completely.

File: src/detect-transform-urldecode.c [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+static void TransformUrlDecode(InspectionBuffer *buffer)
+{
+    const uint8_t *input = buffer->inspect;
+    const uint32_t input_len = buffer->inspect_len;
+    uint8_t output[input_len]; // we can only shrink
+
+    uint32_t output_size = utilBufferUrlDecode(input, input_len, output);
+    printf("lol %d\n", output_size);
```
Comment (victorjulien) [inline src/detect-transform-urldecode.c]: debug leftover

File: doc/userguide/rules/transforms.rst [hunk 5]
```diff
@@ -106,3 +106,7 @@ Example::
 
 .. note:: depends on libnss being compiled into Suricata
 
+url_decode
+-------------------
+
+Decodes url-encoded data, ie replacing '+' with space and '%HH' with its value.
```
Comment (victorjulien) [inline doc/userguide/rules/transforms.rst]: Since the code mentions its not decoding unicode, would it make sense to mention that here as well?

**Comment by victorjulien** [comment]
" Modifies #4347 by fixing transformed buffer size and avoiding overflow "

Idea: libfuzzer interface to each transform?

**Comment by catenacyber** [comment]
>> " Modifies #4347 by fixing transformed buffer size and avoiding overflow "
> Idea: libfuzzer interface to each transform?

Indeed, but it would be interesting to get a sense of coverage of current fuzz targets (oss-fuzz has a nice report for coverage)

File: doc/userguide/rules/transforms.rst (hunk 5 above)
Comment (catenacyber) [reply doc/userguide/rules/transforms.rst]: Ok and I take it as a confirmation that unicode decoding should not be part of this keyword

File: src/detect-http-client-body.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http-client-body.c]: ok

File: src/detect-transform-urldecode.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c]: ok

File: src/detect-transform-urldecode.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c]: ok, sorry for the golang style

File: src/detect-transform-urldecode.c (hunk 4 above)
Comment (catenacyber) [reply src/detect-transform-urldecode.c]: ok
